src/Protocols/Avalon/MemMap/AvalonMemMap.hs

```diff
@@ -0,0 +1,1245 @@
+
+instance Protocol (AvalonMMSubordinate dom fixedWaitTime config) where
+  type Fwd (AvalonMMSubordinate dom fixedWaitTime config)
+         = Signal dom (AvalonSubordinateIn config)
+  type Bwd (AvalonMMSubordinate dom fixedWaitTime config)
+         = Signal dom (AvalonSubordinateOut config)
+
+instance (KnownMMSubordinateConfig config, KeepWaitRequest config ~ 'True) =>
+  Backpressure (AvalonMMSubordinate dom 0 config) where
+
+  boolsToBwd _ = C.fromList_lazy . fmap boolToMMSubordinateAck
+
+instance (KnownMMManagerConfig config) =>
+  Backpressure (AvalonMMManager dom config) where
+
+  boolsToBwd _ = C.fromList_lazy . fmap boolToMMManagerAck
+
+-- 'toDfLike' and 'fromDfLike' here are very similar to their implementations
+-- in @instance DfConv.DfConv (AvalonMMManager ...)@. 'toDfCircuit' forces
+-- burstCount to be 1, will overwrite whatever value you specify in your
+-- WriteImpt and ReadReqImpt. In the case of write, this is because the user
+-- could otherwise send in a write with burst size of 2, but then send a read
+-- request along the same channel, which would either have to get dropped, or
+-- the write burst would have to be left unfinished. In the case of read
+-- request, this is because read data only comes in for one clock cycle, while
+-- a Df acknowledge can take a while to come in, so we need to store incoming
+-- read data; if we were allowed to have arbitrarily large burst sizes, we
+-- would need to store incoming read data in an arbitrarily-large buffer.
+instance
+  (KnownMMSubordinateConfig config, config ~ RemoveNonDfSubordinate config) =>
+  DfConv.DfConv   (AvalonMMSubordinate dom 0 config) where
+
+  type Dom        (AvalonMMSubordinate dom 0 config) = dom
+  type BwdPayload (AvalonMMSubordinate dom 0 config)
+         = AvalonReadImpt (SShared config)
+  type FwdPayload (AvalonMMSubordinate dom 0 config)
+         = Either
+             (AvalonReadReqImpt (KeepAddr config) (SShared config))
+             (AvalonWriteImpt (KeepAddr config) (SShared config))
+
+  toDfCircuit proxy = DfConv.toDfCircuitHelper proxy s0 blankOtp stateFn where
+    s0 = (Nothing, False)
+    -- Nothing: readDatStored:
+    --   reads only get sent for one clock cycle, so we have to store it
+    --   until it's acked
+    -- False: readReqAcked:
+    --   when our read request is acknowledged, we should stop sending the
+    --   read request forwards
+    blankOtp = mmSubordinateInNoData
+    stateFn so dfAck dfDat = do
+      (readDatStored, readReqAcked) <- get
+      let readDatIn = mmSubordinateOutToReadImpt so
+      let miBool = mmSubordinateOutToBool so
+      let (readDatStored', readReqAcked', si, dfAckOut)
+            = case (dfDat, readDatStored) of
+                (Just (Right wi), _) ->
+                  ( readDatStored
+                  , False
+                  , mmWriteImptToSubordinateIn
+                      (wi { wi_burstCount = toKeepType 1 })
+                  , miBool )
```

Comment:
The nesting+wrapping suggests that this `case` might be better of being a function on its own.